### PR TITLE
Implement connection management, gremlin query, execute histroy and g…

### PR DESCRIPTION
**Pull request by Linary**
…remlin collection

### CODE REVIEW

File: hubble-be/src/main/java/com/baidu/hugegraph/controller/SettingController.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                        "The acceptable languages are %s, but got %s",
+                        Constant.LANGUAGES, lang);
+
+        UserInfo userInfo = UserInfo.builder()
+                                    .username("anonymous")
+                                    .locale(lang)
+                                    .build();
+        int rows = this.service.update(userInfo);
+        if (rows != 1) {
+            throw new RuntimeException("Save failed");
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/controller/SettingController.java]: translate

File: hubble-be/src/main/java/com/baidu/hugegraph/controller/SettingController.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+public class SettingController {
+
+    @Autowired
+    private UserInfoService service;
+
+    /**
+     * TODO: To be improved
+     */
+    @GetMapping("common")
+    public UserInfo common(@RequestParam(value = "lang",
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/controller/SettingController.java]: rename to config()

File: hubble-be/src/main/java/com/baidu/hugegraph/controller/SettingController.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                                         defaultValue = "en_US")
+                           String lang,
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/controller/SettingController.java]: add RequestParam style

File: hubble-be/src/main/java/com/baidu/hugegraph/entity/AdjacentQuery.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    private Object vertexId;
+    private String edgeLabel;
+    private Direction direction;
+    private List<Term> terms;
+
+    @Data
+    @NoArgsConstructor
+    @AllArgsConstructor
+    @Builder
+    public static class Term {
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/entity/AdjacentQuery.java]: Term => Condition

File: hubble-be/src/main/java/com/baidu/hugegraph/service/GraphConnectionService.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+    @Autowired
+    private GraphConnectionMapper mapper;
+
+    public List<GraphConnection> listAll() {
+        return this.mapper.selectList(null);
+    }
+
+    public IPage<GraphConnection> list(String content,
+                                       Boolean graphOrderAsc,
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/service/GraphConnectionService.java]: boolean

File: hubble-be/src/main/java/com/baidu/hugegraph/service/GremlinCollectionService.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+import com.baomidou.mybatisplus.extension.plugins.pagination.Page;
+
+@Service
+public class GremlinCollectionService {
+
+    @Autowired
+    private GremlinCollectionMapper mapper;
+
+    public IPage<GremlinCollection> list(String content,
+                                         Boolean nameOrderAsc,
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/service/GremlinCollectionService.java]: ditto

File: hubble-be/src/main/java/com/baidu/hugegraph/util/EntityUtil.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+                    } else {
+                        field.set(entity, newFieldValue);
+                    }
+                } else {
+                    field.set(entity, oldFieldValue);
+                }
+            } catch (IllegalAccessException e) {
+                throw new InternalException(
+                          "Failed to access field %s of class %s",
+                          field.getName(), clazz.getName());
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/util/EntityUtil.java]: need to translate

File: hubble-be/src/main/resources/i18n/messages_en_US.properties [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+common.param.should-belong-to=The param {0} should belong to {1}
+common.param.path-id-should-same-as-body=The id in path({0}) must be same as request body({0}) if it exists
+common.unknown.enum.type=Unknown type {0} for enum {1}
+
+entity.insert.failed=Insert entity {0} failed
+entity.update.failed=Update entity {0} failed
+entity.delete.failed=Delete entity {0} failed
+
+client-server.version.unmatched=The version of hugegraph-client and hugegraph-server don't match
+service.unavailable=The service {0}:{1} is unavailable
```
[the remark this answers is not in the document]
Comment (Linary) [reply hubble-be/src/main/resources/i18n/messages_en_US.properties]: I will modify it in subsequent commit

File: hubble-be/src/main/java/com/baidu/hugegraph/controller/BaseController.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+import com.baidu.hugegraph.util.EntityUtil;
+import com.baidu.hugegraph.util.Ex;
+
+public abstract class BaseController {
+
+    public static final String ORDER_ASC = "asc";
+    public static final String ORDER_DESC = "desc";
+
+    public void checkIdWhenUpdate(int id, Identifiable newEntity) {
+        Ex.check(newEntity.getId() != null, () -> id == newEntity.getId(),
```
[the remark this answers is not in the document]
Comment (Linary) [reply hubble-be/src/main/java/com/baidu/hugegraph/controller/BaseController.java]: I have added comment on method

File: hubble-be/src/main/java/com/baidu/hugegraph/controller/GraphConnectionController.java [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+        this.checkParamsNotEmpty("graph", graph, creating);
+        Ex.check(graph != null, () -> graph.matches(GRAPH_REGEX),
+                 "graph-connection.graph.unmatch-regex", graph);
+
+        this.checkParamsNotEmpty("host", newEntity.getHost(), creating);
+
+        Integer port = newEntity.getPort();
+        Ex.check(creating, () -> port != null,
+                 "common.param.cannot-be-null", "port");
+        Ex.check(port != null, () -> 1 <= port && port <= 65535,
```
[the remark this answers is not in the document]
Comment (Linary) [reply hubble-be/src/main/java/com/baidu/hugegraph/controller/GraphConnectionController.java]: guava has UnsignedInts,  UnsignedLongs, but doesn't have UnsignedShorts

File: hubble-be/src/main/java/com/baidu/hugegraph/controller/SettingController.java (hunk 3 above)
Comment (Linary) [reply hubble-be/src/main/java/com/baidu/hugegraph/controller/SettingController.java]: what means?

File: hubble-be/src/main/java/com/baidu/hugegraph/service/GraphConnectionService.java (hunk 5 above)
Comment (Linary) [reply hubble-be/src/main/java/com/baidu/hugegraph/service/GraphConnectionService.java]: I need if (xxx == null)

File: LICENSE [hunk 11]
```diff
@@ -6,7 +6,7 @@ Apache License
 
    1. Definitions.
 
-      "License" shall mean the terms and conditions for use, reproduction,
+      "License" shall mean the conditions and conditions for use, reproduction,
```
Comment (javeme) [inline LICENSE]: keep the origin

File: hubble-be/assembly/descriptor/assembly.xml [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+            <unpack>false</unpack>
+            <scope>runtime</scope>
+            <useProjectArtifact>false</useProjectArtifact>
+            <includes>
+                <include>*:*:jar:*</include>
+            </includes>
+        </dependencySet>
+    </dependencySets>
+
+</assembly>
```
Comment (javeme) [inline hubble-be/assembly/descriptor/assembly.xml]: expect new line

File: hubble-be/src/main/java/com/baidu/hugegraph/advisor/ExceptionAdvisor.java [hunk 13]
```diff
@@ -79,18 +79,18 @@ public Response exceptionHandler(Exception e) {
...
             for (int i = 0; i < args.length; i++) {
                 strArgs[i] = args[i].toString();
             }
         }
-        String msg = messageKey;
+        String msg = message;
         try {
-            msg = this.messageSourceHandler.getMessage(messageKey, strArgs);
-        } catch (Exception e) {
+            msg = this.messageSourceHandler.getMessage(message, strArgs);
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/advisor/ExceptionAdvisor.java]: message = ...

File: hubble-be/src/main/java/com/baidu/hugegraph/controller/SettingController.java [hunk 14]
```diff
@@ -40,28 +42,39 @@
...
+            int rows = this.service.update(userInfo);
+            if (rows == 0) {
+                throw new InternalException("entity.update.failed");
+            }
         }
-        Cookie cookie = new Cookie("user", userInfo.getUsername());
+
+        Cookie cookie = new Cookie(Constant.COOKIE_USER, userInfo.getUsername());
+        cookie.setPath(request.getContextPath());
+        cookie.setMaxAge(3 * 24 * 60 * 60);
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/controller/SettingController.java]: use TimeUnit

File: hubble-be/src/main/java/com/baidu/hugegraph/service/GraphConnectionService.java [hunk 15]
```diff
@@ -44,10 +44,8 @@
         return this.mapper.selectList(null);
     }
 
-    public IPage<GraphConnection> list(String content,
-                                       Boolean graphOrderAsc,
-                                       long current,
-                                       long pageSize) {
+    public IPage<GraphConnection> list(String content, Boolean graphOrderAsc,
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/service/GraphConnectionService.java]: add some comments

File: hubble-be/src/main/java/com/baidu/hugegraph/service/GremlinQueryService.java [hunk 16]
```diff
@@ -140,7 +140,7 @@ public GremlinResult expandVertex(AdjacentQuery query) {
     }
 
     @SuppressWarnings("unchecked")
-    @Cacheable(value = "explain", key = "#query.content")
+    @Cacheable(value = "GREMLIN_QUERY_EXPLAIN", key = "#query.content")
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/service/GremlinQueryService.java]: error value

File: LICENSE [hunk 17]
```diff
@@ -121,18 +121,18 @@ Apache License
...
 
    5. Submission of Contributions. Unless You explicitly state otherwise,
       any Contribution intentionally submitted for inclusion in the Work
-      by You to the Licensor shall be under the terms and conditions of
-      this License, without any additional terms or conditions.
+      by You to the Licensor shall be under the conditions and conditions of
+      this License, without any additional conditions or conditions.
       Notwithstanding the above, nothing herein shall supersede or modify
-      the terms of any separate license agreement you may have executed
+      the conditions of any separate license agreement you may have executed
```
Comment (javeme) [inline LICENSE]: 6 "conditions" in this file

File: hubble-be/pom.xml [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+            <!--                                <include>${final.name}/**</include>-->
+            <!--                            </includes>-->
+            <!--                            <followSymlinks>false</followSymlinks>-->
+            <!--                        </fileset>-->
+            <!--                        <fileset>-->
+            <!--                            <directory>${final.name}</directory>-->
+            <!--                        </fileset>-->
+            <!--                    </filesets>-->
+            <!--                </configuration>-->
+            <!--            </plugin>-->
```
Comment (javeme) [inline hubble-be/pom.xml]: unused?

File: hubble-be/src/main/java/com/baidu/hugegraph/config/SwaggerConfig.java [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+
+import springfox.documentation.builders.ApiInfoBuilder;
+import springfox.documentation.builders.PathSelectors;
+import springfox.documentation.builders.RequestHandlerSelectors;
+import springfox.documentation.service.ApiInfo;
+import springfox.documentation.spi.DocumentationType;
+import springfox.documentation.spring.web.plugins.Docket;
+
+@Configuration
+public class SwaggerConfig {
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/config/SwaggerConfig.java]: comment

File: hubble-be/src/main/java/com/baidu/hugegraph/service/GremlinQueryService.java [hunk 20]
```diff
@@ -180,61 +144,77 @@ private String buildGremlinQuery(AdjacentQuery query) {
...
+        }
+
+        if (!edges.isEmpty()) {
+            if (vertices.isEmpty()) {
                 vertices = this.verticesOfEdge(edges);
-                break;
-            case PATH:
-                vertices = this.verticesOfPath((List<Path>) (Object) typedData);
+            } else {
+                vertices.putAll(this.verticesOfEdge(edges));
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/service/GremlinQueryService.java]: TODO improve perf

File: hubble-be/src/main/java/com/baidu/hugegraph/service/GremlinQueryService.java [hunk 21]
```diff
@@ -126,42 +129,32 @@ public GremlinResult expandVertex(AdjacentQuery query) {
...
+            } else if (object instanceof Edge) {
+                return Type.EDGE;
+            } else if (object instanceof Path) {
+                return Type.PATH;
+            } else {
+                return Type.GENERAL;
             }
-            sb.append(")");
+        } else {
+            return Type.EMPTY;
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/service/GremlinQueryService.java]: return early

File: hubble-dist/pom.xml [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+                        </fileset>
+                        <fileset>
+                            <directory>${final.name}</directory>
+                        </fileset>
+                    </filesets>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+</project>
```
Comment (javeme) [inline hubble-dist/pom.xml]: expect end line

File: hubble-be/src/main/java/com/baidu/hugegraph/service/GremlinQueryService.java [hunk 23]
```diff
@@ -126,42 +129,32 @@ public GremlinResult expandVertex(AdjacentQuery query) {
...
-        // limit
-        sb.append(".limit(").append(MAX_EDGES_PER_VERTEX).append(")");
-        // other vertex
-        sb.append(".otherV().path()");
-        return sb.toString();
     }
 
+    /**
+     * TODO: reduce the number of requests
+     */
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/service/GremlinQueryService.java]: move to line 200

File: hubble-be/src/main/java/com/baidu/hugegraph/util/EntityUtil.java [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+        T entity;
+        try {
+            entity = (T) clazz.newInstance();
+        } catch (InstantiationException | IllegalAccessException e) {
+            throw new InternalException("reflect.new-instance.failed",
+                                        clazz.getName());
+        }
+        Field[] fields = clazz.getDeclaredFields();
+        for (Field field : fields) {
+            // NOTE: Skip jacoco injected filed
```
Comment (zhoney) [inline hubble-be/src/main/java/com/baidu/hugegraph/util/EntityUtil.java]: field

File: hubble-be/src/main/resources/hugegraph-hubble.properties [hunk 25]
```diff
@@ -0,0 +1 @@
+server.port=8181
```
Comment (zhoney) [inline hubble-be/src/main/resources/hugegraph-hubble.properties]: set to default value like:8088

File: hubble-be/src/main/java/com/baidu/hugegraph/service/GremlinQueryService.java [hunk 26]
```diff
@@ -0,0 +1,292 @@
...
+        if (query.getConditions() != null) {
+            // properties
+            for (AdjacentQuery.Condition condition : query.getConditions()) {
+                // key
+                sb.append(".has('").append(condition.getKey()).append("', ");
+                // value
+                sb.append(condition.getOperator()).append("(")
+                  .append(GremlinUtil.escape(condition.getValue())).append(")");
+            }
+            sb.append(")");
```
[the remark this answers is not in the document]
Comment (Linary) [reply hubble-be/src/main/java/com/baidu/hugegraph/service/GremlinQueryService.java]: LGTM!

File: hubble-be/src/main/java/com/baidu/hugegraph/controller/GremlinCollectionController.java [hunk 27]
```diff
@@ -0,0 +1,163 @@
...
+import com.baidu.hugegraph.exception.InternalException;
+import com.baidu.hugegraph.service.GremlinCollectionService;
+import com.baidu.hugegraph.util.Ex;
+import com.baomidou.mybatisplus.core.metadata.IPage;
+
+@RestController
+@RequestMapping("gremlin-collections")
+public class GremlinCollectionController extends BaseController {
+
+    private static final String NAME_REGEX = "[A-Za-z0-9_]{1,50}";
```
[the remark this answers is not in the document]
Comment (Linary) [reply hubble-be/src/main/java/com/baidu/hugegraph/controller/GremlinCollectionController.java]: prd doc required

**Pull request merged**